src/raw_client.rs

```diff
@@ -668,32 +667,32 @@ impl<T: Read + Write> ElectrumApi for RawClient<T> {
 
         self.increment_calls();
 
-        while !missing_responses.is_empty() {
-            let req_id = *missing_responses.iter().next().unwrap();
-            let resp = match self.recv(&receiver, req_id) {
-                Ok(resp) => resp,
+        let mut answers = BTreeMap::new();
+        for req_id in missing_responses {
+            match self.recv(&receiver, req_id) {
                 Err(e) => {
                     // In case of error our sender could still be left in the map, depending on where
                     // the error happened. Just in case, try to remove it here
                     warn!("got error for req_id {}: {:?}", req_id, e);
                     warn!("removing all waiting req of this batch");
-                    let mut guard = self.waiting_map.lock().unwrap();
-                    for req_id in missing_responses.iter() {
-                        guard.remove(&req_id);
-                    }
+                    let mut guard = self
+                        .waiting_map
+                        .lock()
+                        .map_err(|_| Error::CouldntLockReader)?;
+                    guard.remove(&req_id);
```

Comment:
I think I'm missing something, why did you remove the loop here?

Reply:
Yep,my bad, I haven't understood the code and thought `req_id` is an external variable. Will fix. Now I see why you was maintaining list of missing responses up to date.

In general, can you explain what actually should happen here? What should be a consequences of erroring one of the batch request? Cancellation of the rest of the requests and returning single error, ignoring all ok-requests already received? That does not make a lot of sense for me... May be I am missing something?

Reply:
Right now if one request fails it makes the whole batch fail. If there are still some requests pending it will just ignore their response when they come back (there's no way of cancelling a request you've already sent).

This line specifically is where all the pending requests id are removed from the map of responses we are waiting for, which means that once the response comes in the current `reader_thread` will just ignore it:

https://github.com/bitcoindevkit/rust-electrum-client/blob/02462b37b383ad9787ad95f53bcf8b5e5d17f616/src/raw_client.rs#L493-L500